File: mgyoutube-multitech/api-java/src/main/java/com/djpedesen/mgyoutube/api_java/webservices/ParentsWebService.java
package com.djpedesen.mgyoutube.api_java.webservices;

import java.io.UnsupportedEncodingException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.List;

import javax.ws.rs.Consumes;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.Response.Status;

import com.djpedesen.mgyoutube.api_java.apimodel.User;
import com.djpedesen.mgyoutube.api_java.apimodel.UserCredential;
import com.djpedesen.mgyoutube.api_java.modules.SearchesModule;
import com.djpedesen.mgyoutube.api_java.modules.UserModule;
import com.djpedesen.mgyoutube.api_java.modules.UserNotFoundException;
import com.google.gson.Gson;

@Path("/parents")
public class ParentsWebService {
	private static final Gson GSON = new Gson();

	private UserModule userModule;
	private SearchesModule searchesModule;

	public ParentsWebService() {
		this.userModule = ModuleRepoRegistry.getUserModule();
		this.searchesModule = ModuleRepoRegistry.getSearchesModule();
	}

	@POST
	@Path("/auth")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	public Response authParentUser(final String userCredentialJson) {
		try {
			System.out.println("authParentUser: userCredentialJson=" + userCredentialJson);

			// TODO: need to sanitize payload input before using
			final String sanitizedUserCredentialJson = userCredentialJson;
			final UserCredential userCredential = Helpers.marshalUserCredentialFromJson(sanitizedUserCredentialJson);
			System.out.println("authParentUser: userCredential=" + userCredential);

			final User authedUser = userModule.authUser(userCredential);

			if (authedUser == null) {
				System.out.println("authParentUser: userCredentialJson=" + sanitizedUserCredentialJson
						+ " failed auth, returning UNAUTHORIZED");
				return Response.status(Status.UNAUTHORIZED).build();
			} else if (!authedUser.isParent) {
				System.out.println("authParentUser: userCredentialJson=" + sanitizedUserCredentialJson
						+ " is not a parent, returning UNAUTHORIZED");
				return Response.status(Status.UNAUTHORIZED).build();
			}

			authedUser.password = null;

			final String responseJson = GSON.toJson(authedUser);

			System.out.println("authParentUser: userCredentialJson=" + sanitizedUserCredentialJson + " returning OK");
			return Response.ok(responseJson, MediaType.APPLICATION_JSON).build();
		} catch (Throwable thrown) {
			System.out.println("authParentUser, unhandled exception:");
			thrown.printStackTrace(System.out);
			return Response.serverError().build();
		}
	}

	@GET
	@Path("/{parentUsername}/children")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getChildrenForParent(@PathParam(value = "parentUsername") final String parentUsername) {
		try {
			System.out.println("getChildrenForParent: parentUsername=" + parentUsername);

			// TODO: need to sanitize payload input before using
			final String sanitizedParentUsername = parentUsername;

			try {
				final List<User> children = userModule.getChildrenForParent(sanitizedParentUsername);
				System.out.println("getChildrenForParent: getChildrenForParent=" + children);

				children.forEach(u -> {
					u.password = null;
				});

				final String responseJson = GSON.toJson(children);

				System.out.println("getChildrenForParent: parentUsername=" + sanitizedParentUsername
						+ " returning OK for size " + children.size() + " and json=" + responseJson);
				return Response.ok(responseJson, MediaType.APPLICATION_JSON).build();
			} catch (UserNotFoundException unfe) {
				System.out
						.println("getChildrenForParent: unable to find user " + unfe.username + " returning NOT_FOUND");
				return Response.status(Status.NOT_FOUND).build();
			}
		} catch (Throwable thrown) {
			System.out.println("getChildrenForParent, unhandled exception:");
			thrown.printStackTrace(System.out);
			return Response.serverError().build();
		}
	}

	@PUT
	@Path("/{parentUsername}/children/{childUsername}")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	public Response addUpdateChildToParent(@PathParam(value = "parentUsername") final String parentUsername,
			@PathParam(value = "childUsername") final String childUsername, final String childUserJson) {
		try {
			System.out.println("addUpdateChildToParent: parentUsername=" + parentUsername + ", childUsername="
					+ childUsername + ", childUserJson=" + childUserJson);

			// TODO: need to sanitize payload input before using
			final String sanitizedParentUsername = parentUsername;
			final String sanitizedChildUsername = childUsername;
			final String sanitizedChildUserJson = childUserJson;

			final User childUser = Helpers.marshalUserFromJson(sanitizedChildUserJson);
			System.out.println("addUpdateChildToParent: childUser=" + childUser);

			try {
				final User addedUpdatedUser = userModule.addUpdateChildToParent(sanitizedParentUsername, childUser);

				addedUpdatedUser.password = null;

				final String responseJson = GSON.toJson(addedUpdatedUser);

				System.out.println("addUpdateChildToParent: parentUsername=" + sanitizedParentUsername
						+ ", childUsername=" + sanitizedChildUsername + " returning OK");
				return Response.ok(responseJson, MediaType.APPLICATION_JSON).build();
			} catch (UserNotFoundException unfe) {
				System.out.println(
						"addUpdateChildToParent: unable to find user " + unfe.username + " returning NOT_FOUND");
				return Response.status(Status.NOT_FOUND).build();
			}
		} catch (Throwable thrown) {
			System.out.println("addUpdateChildToParent, unhandled exception:");
			thrown.printStackTrace(System.out);
			return Response.serverError().build();
		}
	}

	@GET
	@Path("/{parentUsername}/searches")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getSavedSearches(@PathParam(value = "parentUsername") final String parentUsername) {
		try {
			System.out.println("getSavedSearches: parentUsername=" + parentUsername);

			// TODO: need to sanitize payload input before using
			final String sanitizedParentUsername = parentUsername;

			try {
				final List<String> searches = searchesModule.getSavedSearchesForParent(sanitizedParentUsername);
				System.out.println("getSavedSearches: getSavedSearchesForParent=" + searches);

				final String responseJson = GSON.toJson(searches);

				System.out.println("getSavedSearches: parentUsername=" + sanitizedParentUsername
						+ " returning OK for size " + searches.size() + " and json=" + responseJson);
				return Response.ok(responseJson, MediaType.APPLICATION_JSON).build();
			} catch (UserNotFoundException unfe) {
				System.out.println("getSavedSearches: unable to find user " + unfe.username + " returning NOT_FOUND");
				return Response.status(Status.NOT_FOUND).build();
			}
		} catch (Throwable thrown) {
			System.out.println("getSavedSearches, unhandled exception:");
			thrown.printStackTrace(System.out);
			return Response.serverError().build();
		}
	}

	@PUT
	@Path("/{parentUsername}/searches/{searchphrase}")
	public Response saveSearch(@PathParam(value = "parentUsername") final String parentUsername,
			@PathParam(value = "searchphrase") final String searchPhrase) throws UnsupportedEncodingException {
		try {
			final String decodedSearchPhrase = URLDecoder.decode(searchPhrase, StandardCharsets.UTF_8.toString());
			System.out.println("saveSearch: parentUsername=" + parentUsername + ", searchPhrase=" + searchPhrase
					+ ", decodedSearchPhrase=" + decodedSearchPhrase);

			// TODO: need to sanitize payload input before using
			final String sanitizedParentUsername = parentUsername;
			final String sanitizedSearchPhrase = decodedSearchPhrase;

			try {
				searchesModule.addSearchToParent(sanitizedParentUsername, sanitizedSearchPhrase);

				System.out.println("saveSearch: parentUsername=" + sanitizedParentUsername + ", searchPhrase="
						+ sanitizedSearchPhrase + " returning CREATED");
				final URI location = new URI("/parents/" + sanitizedParentUsername + "/searches");
				return Response.created(location).build();
			} catch (UserNotFoundException unfe) {
				System.out.println("saveSearch: unable to find user " + unfe.username + " returning NOT_FOUND");
				return Response.status(Status.NOT_FOUND).build();
			} catch (URISyntaxException use) {
				System.out.println("saveSearch: problem creating location for response");
				use.printStackTrace(System.out);
				return Response.serverError().build();
			}
		} catch (Throwable thrown) {
			System.out.println("saveSearch, unhandled exception:");
			thrown.printStackTrace(System.out);
			return Response.serverError().build();
		}
	}

}

File: mgyoutube-multitech/api-python/webservices/ParentsWebService.py
import json
import urllib.parse
from pyramid.response import Response
from pyramid.request import Request
from pyramid.view import view_config
from pyramid.config import Configurator
from pyramid.httpexceptions import HTTPNotFound, HTTPUnauthorized, HTTPMethodNotAllowed,  HTTPServerError
from webservices.ModuleRepoRegistry import ModuleRepoRegistry
from webservices.Helpers import Helpers
from modules.UserNotFoundException import UserNotFoundException
from modules.UserModule import UserModule


@view_config(route_name='parents-auth', request_method='POST')
def authParentUser(request: Request) -> Response:
    userCredentialJson = request.body
    print("authParentUser: userCredentialJson=",  userCredentialJson)

    # TODO: need to sanitize payload input before using
    sanitizedUserCredentialJson = userCredentialJson
    userCredential = Helpers.marshalUserCredentialFromJson(
        sanitizedUserCredentialJson)
    print("authParentUser: userCredential=", userCredential)

    authedUser = ModuleRepoRegistry.getUserModule().authUser(userCredential)

    if (authedUser is None):
        print("authParentUser: userCredentialJson=", userCredentialJson,
              " failed auth, returning UNAUTHORIZED")
        raise HTTPUnauthorized()
    elif (not authedUser.isParent):
        print("authParentUser: userCredentialJson=", userCredentialJson,
              " is not a parent, returning UNAUTHORIZED")
        raise HTTPUnauthorized()

    authedUser.password = None
    responseJson = json.dumps(authedUser.__dict__)

    print("authParentUser: userCredentialJson=",
          userCredentialJson, " returning OK")
    return Response(body=responseJson, content_type='application/json', charset="UTF-8")


@view_config(route_name='parents-children', request_method='GET')
def getChildrenForParent(request: Request) -> Response:

    # TODO: need to sanitize payload input before using
    parentUsername = request.matchdict['parentusername']
    sanitizedParentUsername = parentUsername
    print("getChildrenForParent: parentUsername=", parentUsername)

    try:
        children = ModuleRepoRegistry.getUserModule(
        ).getChildrenForParent(sanitizedParentUsername)
        print("getChildrenForParent: getChildrenForParent", children)

        for u in children:
            u.password = None

        responseJson = json.dumps(children, default=lambda x: x.__dict__)

        print("getChildrenForParent: parentUsername", parentUsername,
              " returning OK for size ", len(children), " and json ", responseJson)
        return Response(body=responseJson, content_type='application/json', charset="UTF-8")
    except UserNotFoundException as unfe:
        print("getChildrenForParent: unable to find user ",
              unfe.username, " returning NOT_FOUND")
        raise HTTPNotFound()


@view_config(route_name='parents-addupdatechildren', request_method='PUT')
def addUpdateChildToParent(request: Request) -> Response:

    # TODO: need to sanitize payload input before using
    parentUsername = request.matchdict['parentusername']
    sanitizedParentUsername = parentUsername
    childUsername = request.matchdict['childusername']
    # sanitizedChildUsername = childUsername
    childUserJson = request.body
    sanitizedChildUserJson = childUserJson
    print("addUpdateChildToParent: parentUsername=", parentUsername,
          ", childUsername=", childUsername, ", childUserJson=", childUserJson)

    childUser = Helpers.marshalUserFromJson(sanitizedChildUserJson)
    print("addUpdateChildToParent: childUser=", childUser)

    try:
        addedUpdatedUser = ModuleRepoRegistry.getUserModule().addUpdateChildToParent(
            sanitizedParentUsername, childUser)

        addedUpdatedUser.password = None

        responseJson = json.dumps(addedUpdatedUser.__dict__)

        print("addUpdateChildToParent: parentUsername=", parentUsername,
              ", childUsername=", childUsername, " returning OK")
        return Response(body=responseJson, content_type='application/json', charset="UTF-8")
    except UserNotFoundException as unfe:
        print("addUpdateChildToParent: unable to find user ",
              unfe.username, " returning NOT_FOUND")
        raise HTTPNotFound()


@view_config(route_name='parents-searches', request_method='GET')
def getSavedSearches(request: Request) -> Response:

    parentUsername = request.matchdict['parentusername']
    print("getSavedSearches: parentUsername=", parentUsername)

    # // TODO: need to sanitize payload input before using
    sanitizedParentUsername = parentUsername

    try:
        searches = ModuleRepoRegistry.getSearchesModule().getSavedSearchesForParent(
            sanitizedParentUsername)
        print("getSavedSearches: getSavedSearchesForParent=", searches)

        responseJson = json.dumps(searches, default=lambda x: x.__dict__)

        print("getSavedSearches: parentUsername=", parentUsername,
              " returning OK for size ", len(searches), " and json=", responseJson)
        return Response(body=responseJson, content_type='application/json', charset="UTF-8")
    except UserNotFoundException as unfe:
        print("getSavedSearches: unable to find user ",
              unfe.username, " returning NOT_FOUND")
        raise HTTPNotFound()


@view_config(route_name='parents-addsearchphrase', request_method='PUT')
def saveSearch(request: Request) -> Response:
    parentUsername = request.matchdict['parentusername']
    searchPhrase = request.matchdict['searchphrase']
    decodedSearchPhrase = urllib.parse.unquote_plus(searchPhrase)
    print("saveSearch: parentUsername=", parentUsername, ", searchPhrase=",
          searchPhrase, ", decodedSearchPhrase=", decodedSearchPhrase)

    # // TODO: need to sanitize payload input before using
    sanitizedParentUsername = parentUsername
    sanitizedSearchPhrase = decodedSearchPhrase

    try:
        ModuleRepoRegistry.getSearchesModule().addSearchToParent(
            sanitizedParentUsername, sanitizedSearchPhrase)

        print("saveSearch: parentUsername=", sanitizedParentUsername,
              ", searchPhrase=", sanitizedSearchPhrase, " returning CREATED")
        location = "/parents/" + sanitizedParentUsername + "/searches"
        return Response(status=201, location=location)
    except UserNotFoundException as unfe:
        print("saveSearch: unable to find user ",
              unfe.username, " returning NOT_FOUND")
        return HTTPNotFound()


def setupParentsWebService(config: Configurator):
    config.add_route('parents-auth',        '/api/parents/auth')
    config.add_route('parents-children',
                     '/api/parents/{parentusername}/children')
    config.add_route('parents-addupdatechildren',
                     '/api/parents/{parentusername}/children/{childusername}')
    config.add_route('parents-searches',
                     '/api/parents/{parentusername}/searches')
    config.add_route('parents-addsearchphrase',
                     '/api/parents/{parentusername}/searches/{searchphrase}')
